File: leetcode/src/validSudoku.cpp
/*
 * determine if a Sudoku is valid. null cell is char '.'
 * */

#include <cstring>
#include <vector>

using namespace std;

class Solution{

public:
    /*
     * @param board is 9*9
     * */
    bool isValidSudoku(const vector<vector<char> > &board){
        const int n = board.size();

        vector<char> block;

        /*
         * check each row
         * */
        for(int i = 0; i < n; ++i){
            block.clear();

            for(int j = 0; j < n; ++j){
                char ch = board[i][j];
                if(ch != '.')    block.push_back(ch);
            }

            if(!check(block))    return false;
        }

        /*
         * check each column
         * */
        for(int j = 0; j < n; ++j){
            block.clear();

            for(int i = 0; i < n; ++i){
                char ch = board[i][j];
                if(ch != '.')    block.push_back(ch);
            }

            if(!check(block))    return false;
        }

        /*
         * check every 3*3 cube
         * */
        int cnt = 9;
        block.clear();

        // tricky to map every 9 cursor numbers to a 3*3 cube
        for (int t = 0; t < 81; t++){
            int l = t / 9; // l remains constant for every 9 numbers, and vary in [0, 8] in total
            int r = t % 9; // r is [0, 8] for every continuous 9 numbers

            // [cx][cy] is left up corner of the 3*3 cube that the int belongs to
            int cx = l / 3 * 3;  // f(x) = x / 3 * 3 is to floor x to multiplies of 3, like [0,3,6]
            int cy = l % 3 * 3;  // f(x) = x % 3 * 3 is to map every [0,1,2] to [0,3,6]

            char ch = board[cx + r / 3][cy + r % 3];  // real position on board

            if(ch != '.')    block.push_back(ch);

            cnt--;

            if(cnt == 0){
                if(!check(block))    return false;

                block.clear();
                cnt += 9;
            }
        }

        return true;
    }

private:
    /*
     * check a vector<> whether have < 9 or > 9 or duplicate int
     * */
    bool check(const vector<char>& vec){
        int pos[9];
        memset(pos, 0, sizeof(pos));

        const int n = vec.size();
        for(int i = 0; i < n; i++){
            int offset = vec[i] - '1';

            if(pos[offset] == 1 || offset < 0 || offset > 8)    return false;

            pos[offset] = 1;
        }

        return true;
    }
};

/* unit test is in ../cpp_unittest/validSudoku_unittest */

File: leetcode/java/validSudoku.java
/*
 * determine if a (9*9) Sudoku is valid. the empty cell is '.' 
 * */
import java.io.*;
import java.util.*;

public class validSudoku{
    public void displayArray(char[] arr){
        for(char ch : arr){
            System.out.print(ch + " ");
        }
        System.out.println();
    }

    private boolean check(char[] arr){
        if(arr.length != 9)    return false;
        HashSet<Integer> hs = new HashSet<Integer>();
        for(char ch : arr){
            if(ch == '.')    continue;
            int d = Integer.parseInt(new Character(ch).toString());
            if(d<0 || d>9 || hs.contains(d))    return false;
            hs.add(d);
        }
        return true;
    }

    public boolean isValidSudoku(char[][] board){
        int n = board.length;
        if(n != 9)    return false;
        char[] block = new char[9];
        for(int i=0;i<9;i++){    //rows
            int c = 0;
            for(int j=0;j<9;j++){
                block[c++] = board[i][j];
            }
            if(!check(block))    return false;
        }
        
        for(int j=0;j<9;++j){  //columns
            int c=0;
            for(int i=0;i<9;++i){
                block[c++] = board[i][j];
            }
            if(!check(block))    return false;
        }

        for(int i=0,j=0; i<=6; j+=3){  //3*3 squares, in rows of 3 cells
            if(j>6){
                j=0;
                i+=3;
            }
            if(i>6)        break;
            int c=0;
            for(int t=i, k=j; t<=2+i; ++k){
                if(k > 2+j){
                    k=j;
                    ++t;
                }
                if(t > 2+i)    break;
                block[c++] = board[t][k];
            }
            if(!check(block))    return false;
        }
        return true;
    }

    public void test_01(){
        char[][] board = new char[9][9];
        char[] arr = new char[]{
            '.', '8', '7', '6', '5', '4', '3', '2', '1'
           ,'2', '.', '.', '.', '.', '.', '.', '.', '.'
           ,'3', '.', '.', '.', '.', '.', '.', '.', '.'
           ,'4', '.', '.', '.', '.', '.', '.', '.', '.'
           ,'5', '.', '.', '.', '.', '.', '.', '.', '.'
           ,'6', '.', '.', '.', '.', '.', '.', '.', '.'
           ,'7', '.', '.', '.', '.', '.', '.', '.', '.'
           ,'8', '.', '.', '.', '.', '.', '.', '.', '.'
           ,'9', '.', '.', '.', '.', '.', '.', '.', '.'
        };
        for(int i=0;i<81;i++){
            board[i/9][i%9] = arr[i];
        }
        System.out.println(isValidSudoku(board));
    }
    
    public static void main(String[] args){
        validSudoku vs = new validSudoku();
        vs.test_01();
        return;
    }
}
